wordpress-mobile/WordPress-iOS
# Create sites with user's timezone

Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/16231

Related:
- https://github.com/wordpress-mobile/WordPressKit-iOS/pull/481

Sites created using the WPiOS app were given a timezone of UTC. The app was also displaying dates and times in UTC, such as when scheduling a post for later publishing, which led to user confusion.

For example, if the user is in UTC-3:00 and opens the post scheduling screen at 19:53 local time, the screen would (incorrectly) show 22:53 in the time picker. If the user wished to schedule the post for 30 minutes later, they might choose 20:23 in the date picker. What would then happen is that the post would publish immediately because the user unknowingly chose 20:23 UTC (which has already passed, given the current time is 22:53 UTC). This would obviously be a source of user confusion since the app wasn't behaving intuitively.

This PR uses `TimeZone.autoupdatingCurrent.identifier` to get the current device time zone. It's my understanding the string returned by this uses the [TZ database format](https://en.wikipedia.org/wiki/Tz_database) e.g. "Pacific/Samoa", which appears to be the standard time zone string format on both iOS and WordPress.

## To test

1. Create a new site using WPiOS app
2. Verify that the timezone of the newly created site matches the device's timezone (look under My Site → Site Settings → Time Zone)
3. Verify that the scheduled post functionality still works as expected
  a. Create a new post
  b. Schedule it for the future (e.g. set the publish date to be now + 2 minutes) under Post Settings
  c. ⚠️ Important: Due to https://github.com/wordpress-mobile/WordPress-iOS/issues/10456, you will have to visit the site once for post scheduling to work
  d. When the scheduled publish time passes, verify that the post is now published
4. Open the WP.com dashboard, choose the site created in Step 1, verify that time zones appear as expected (in both Settings → General → "Site timezone" and also in the post scheduling date picker found inside the block editor)
5. Optional: Change your device time zone and repeat steps 3 and 4.

## Regression Notes

#### Potential unintended areas of impact

We should watch out for site creation regressions, issues with other time zones, and compatibility issues with the web.

#### What I did to test those areas of impact (or what existing automated tests I relied on)

I tested a site created with these changes and made sure post scheduling (which relies heavily on time zones) works correctly. I tested using the `America/Asuncion` timezone to ensure it worked correctly. I then verified that the site I created from WPiOS worked correctly when opened via WP.com.

#### What automated tests I added (or what prevented me from doing so)

I added an automated test in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/481 to ensure the `timezone_string` is correctly encoded in site creation requests. I'm not aware of any suitable tests to add to this repo, but welcome any ideas!

PR submission checklist:
